diff -rupbN linux.orig/arch/arm64/boot/dts/qcom/sdm845-ayn-odin.dts linux/arch/arm64/boot/dts/qcom/sdm845-ayn-odin.dts
--- linux.orig/arch/arm64/boot/dts/qcom/sdm845-ayn-odin.dts	2022-06-13 16:08:51.000000000 +0000
+++ linux/arch/arm64/boot/dts/qcom/sdm845-ayn-odin.dts	2025-07-08 22:59:39.981768534 +0000
@@ -5,16 +5,18 @@
 
 /dts-v1/;
 
+#include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/pinctrl/qcom,pmic-gpio.h>
 #include <dt-bindings/regulator/qcom,rpmh-regulator.h>
 #include <dt-bindings/sound/qcom,q6afe.h>
 #include <dt-bindings/sound/qcom,q6asm.h>
+#include <dt-bindings/leds/common.h>
 #include "sdm845.dtsi"
 #include "pm8998.dtsi"
 #include "pmi8998.dtsi"
 
 / {
-	model = "Odin - AYN Technologies Co., Ltd.";
+	model = "AYN Odin";
 	compatible = "ayn,odin", "qcom,sdm845", "qcom,mtp";
 	qcom,msm-id = <321 0x20001>;
 	qcom,board-id = <8 0>;
@@ -31,7 +33,6 @@
 
 	gpio-keys {
 		compatible = "gpio-keys";
-		autorepeat;
 
 		pinctrl-names = "default";
 		pinctrl-0 = <&volume_up_gpio>;
@@ -43,6 +44,268 @@
 		};
 	};
 
+	gamepad {
+		compatible = "ayntec,odin-gamepad";
+
+		dpad-down-gpio = <&tlmm 128 GPIO_ACTIVE_LOW>;
+		dpad-left-gpio = <&tlmm 52 GPIO_ACTIVE_LOW>;
+		dpad-right-gpio = <&tlmm 122 GPIO_ACTIVE_LOW>;
+		dpad-up-gpio = <&tlmm 124 GPIO_ACTIVE_LOW>;
+		east-btn-gpio = <&tlmm 96 GPIO_ACTIVE_LOW>;
+		enable-gpio = <&tlmm 88 GPIO_ACTIVE_HIGH>;
+		l1-btn-gpio = <&tlmm 26 GPIO_ACTIVE_LOW>;
+		home-btn-gpio = <&tlmm 49 GPIO_ACTIVE_LOW>;
+		north-btn-gpio = <&tlmm 92 GPIO_ACTIVE_LOW>;
+		r1-btn-gpio = <&tlmm 91 GPIO_ACTIVE_LOW>;
+		rear-l-btn-gpio = <&tlmm 116 GPIO_ACTIVE_LOW>;
+		rear-r-btn-gpio = <&tlmm 24 GPIO_ACTIVE_LOW>;
+		select-btn-gpio = <&tlmm 0 GPIO_ACTIVE_LOW>;
+		south-btn-gpio = <&tlmm 95 GPIO_ACTIVE_LOW>;
+		start-btn-gpio = <&tlmm 3 GPIO_ACTIVE_LOW>;
+		thumb-l-btn-gpio = <&tlmm 40 GPIO_ACTIVE_LOW>;
+		thumb-r-rtn-gpio = <&tlmm 22 GPIO_ACTIVE_LOW>;
+		west-btn-gpio = <&tlmm 41 GPIO_ACTIVE_LOW>;
+
+		io-channel-names = "x-axis", "y-axis", "rx-axis",
+				   "ry-axis", "l2-trigger", "r2-trigger";
+		io-channels = <&pm8998_adc ADC5_GPIO3>,
+			      <&pm8998_adc ADC5_GPIO2>,
+			      <&pm8998_adc ADC5_GPIO5>,
+			      <&pm8998_adc ADC5_GPIO6_DIV3>,
+			      <&pm8998_adc ADC5_GPIO1>,
+			      <&pm8998_adc ADC5_GPIO4>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&btn_pins &btn_en_active>;
+
+		x-axis {
+			abs-range = <1110000>;
+			abs-flat = <4500>;
+			abs-fuzz = <8500>;
+		};
+
+		y-axis {
+			abs-range = <1340000>;
+			inverted;
+			abs-flat = <4500>;
+			abs-fuzz = <8500>;
+		};
+
+		rx-axis {
+			abs-range = <1220000>;
+			inverted;
+			abs-flat = <4500>;
+			abs-fuzz = <8500>;
+		};
+
+		ry-axis {
+			abs-range = <1180000>;
+			inverted;
+			abs-flat = <4500>;
+			abs-fuzz = <8500>;
+		};
+
+		l2-trigger {
+			abs-range = <1400000>;
+			inverted;
+			abs-flat = <4500>;
+			abs-fuzz = <8500>;
+		};
+
+		r2-trigger {
+			abs-range = <1400000>;
+			inverted;
+			abs-flat = <4500>;
+			abs-fuzz = <8500>;
+		};
+	};
+
+	fan_en_reg: fan-en-regulator {
+		compatible = "regulator-fixed";
+		regulator-name = "fan_en";
+
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+
+		gpio = <&tlmm 44 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+        pwm_fan: pwm-fan {
+                compatible = "pwm-fan";
+                cooling-levels = <0 40 127>;
+                #cooling-cells = <2>;
+                fan-supply = <&fan_en_reg>;
+                pinctrl-0 = <&pwm_fan_ctrl>;
+                pinctrl-names = "default";
+                pwms = <&pmi8998_lpg 1 50000>;
+        };
+
+	thermal-zones {
+		cluster0-thermal {
+			trips {
+				cluster0_fan0: cluster0-fan0 {
+					temperature = <50000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+				cluster0_fan1: cluster0-fan1 {
+					temperature = <60000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+				cluster0_fan2: cluster0-fan2 {
+					temperature = <70000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+			};
+			cooling-maps {
+				map1 {
+					trip = <&cluster0_fan0>;
+					cooling-device = <&pwm_fan THERMAL_NO_LIMIT 1>;
+				};
+				map2 {
+					trip = <&cluster0_fan1>;
+					cooling-device = <&pwm_fan 2 3>;
+				};
+				map3 {
+					trip = <&cluster0_fan2>;
+					cooling-device = <&pwm_fan 3 THERMAL_NO_LIMIT>;
+				};
+			};
+		};
+		cluster1-thermal {
+			trips {
+				cluster1_fan0: cluster1-fan0 {
+					temperature = <50000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+				cluster1_fan1: cluster1-fan1 {
+					temperature = <60000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+				cluster1_fan2: cluster1-fan2 {
+					temperature = <70000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+			};
+			cooling-maps {
+				map1 {
+					trip = <&cluster1_fan0>;
+					cooling-device = <&pwm_fan THERMAL_NO_LIMIT 1>;
+				};
+				map2 {
+					trip = <&cluster1_fan1>;
+					cooling-device = <&pwm_fan 2 3>;
+				};
+				map3 {
+					trip = <&cluster1_fan2>;
+					cooling-device = <&pwm_fan 3 THERMAL_NO_LIMIT>;
+				};
+			};
+		};
+		gpu-top-thermal {
+			trips {
+				gpu_top_fan0: gpu-top-fan0 {
+					temperature = <50000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+				gpu_top_fan1: gpu-top-fan1 {
+					temperature = <60000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+				gpu_top_fan2: gpu-top-fan2 {
+					temperature = <70000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+			};
+			cooling-maps {
+				map1 {
+					trip = <&gpu_top_fan0>;
+					cooling-device = <&pwm_fan THERMAL_NO_LIMIT 1>;
+				};
+				map2 {
+					trip = <&gpu_top_fan1>;
+					cooling-device = <&pwm_fan 2 3>;
+				};
+				map3 {
+					trip = <&gpu_top_fan2>;
+					cooling-device = <&pwm_fan 3 THERMAL_NO_LIMIT>;
+				};
+			};
+		};
+		gpu-bottom-thermal {
+			trips {
+				gpu_bottom_fan0: gpu-bottom-fan0 {
+					temperature = <50000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+				gpu_bottom_fan1: gpu-bottom-fan1 {
+					temperature = <60000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+				gpu_bottom_fan2: gpu-bottom-fan2 {
+					temperature = <70000>;
+					hysteresis = <2000>;
+					type = "active";
+				};
+			};
+			cooling-maps {
+				map1 {
+					trip = <&gpu_bottom_fan0>;
+					cooling-device = <&pwm_fan THERMAL_NO_LIMIT 1>;
+				};
+				map2 {
+					trip = <&gpu_bottom_fan1>;
+					cooling-device = <&pwm_fan 2 3>;
+				};
+				map3 {
+					trip = <&gpu_bottom_fan2>;
+					cooling-device = <&pwm_fan 3 THERMAL_NO_LIMIT>;
+				};
+			};
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&gpio_leds_default>;
+
+		led-0 {
+			color = <LED_COLOR_ID_BLUE>;
+			gpios = <&tlmm 7 GPIO_ACTIVE_HIGH>;
+			label = "right_joystick";
+		};
+
+		led-1 {
+			color = <LED_COLOR_ID_BLUE>;
+			gpios = <&tlmm 25 GPIO_ACTIVE_HIGH>;
+			label = "left_joystick";
+		};
+
+		led-2 {
+			color = <LED_COLOR_ID_BLUE>;
+			gpios = <&tlmm 87 GPIO_ACTIVE_HIGH>;
+			label = "right_side";
+		};
+
+		led-3 {
+			color = <LED_COLOR_ID_BLUE>;
+			gpios = <&tlmm 129 GPIO_ACTIVE_HIGH>;
+			label = "left_side";
+		};
+	};
+
 	battery: battery {
 		compatible = "simple-battery";
 
@@ -74,7 +337,7 @@
 
 &adsp_pas {
 	status = "okay";
-	firmware-name = "qcom/sdm845/odin/adsp.mbn";
+	firmware-name = "qcom/sdm845/adsp.mbn";
 };
 
 &apps_rsc {
@@ -192,6 +455,12 @@
 			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
 		};
 
+		vreg_l23a_3p3: ldo23 {
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3312000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
 		vreg_l24a_3p075: ldo24 {
 			regulator-min-microvolt = <3088000>;
 			regulator-max-microvolt = <3088000>;
@@ -250,7 +519,7 @@
 
 &cdsp_pas {
 	status = "okay";
-	firmware-name = "qcom/sdm845/odin/cdsp.mbn";
+	firmware-name = "qcom/sdm845/cdsp.mbn";
 };
 
 &dsi0 {
@@ -314,15 +583,10 @@
 	status = "okay";
 	zap-shader {
 		memory-region = <&gpu_mem>;
-		firmware-name = "qcom/sdm845/odin/a630_zap.mbn";
+		firmware-name = "qcom/sdm845/a630_zap.mbn";
 	};
 };
 
-&pmi8998_haptics {
-	status = "okay";
-	qcom,wave-play-rate-us = <4255>;
-};
-
 &i2c5 {
 	status = "okay";
 	clock-frequency = <400000>;
@@ -384,7 +648,51 @@
 
 &mss_pil {
 	status = "okay";
-	firmware-name = "qcom/sdm845/odin/mba.mbn", "qcom/sdm845/odin/modem.mbn";
+	firmware-name = "qcom/sdm845/mba.mbn", "qcom/sdm845/modem.mbn";
+};
+
+&pm8998_adc {
+	channel@12 {
+		reg = <ADC5_GPIO1>;
+		label = "gpio1";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	channel@13 {
+		reg = <ADC5_GPIO2>;
+		label = "gpio2";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	channel@14 {
+		reg = <ADC5_GPIO3>;
+		label = "gpio3";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	channel@15 {
+		reg = <ADC5_GPIO4>;
+		label = "gpio4";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	channel@16 {
+		reg = <ADC5_GPIO5>;
+		label = "gpio5";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 1>;
+	};
+
+	channel@97 {
+		reg = <ADC5_GPIO6_DIV3>;
+		label = "gpio6";
+		qcom,decimation = <256>;
+		qcom,pre-scaling = <1 3>;
+	};
 };
 
 &pm8998_gpio {
@@ -407,6 +715,11 @@
 	};
 };
 
+&pmi8998_charger {
+	status = "okay";
+	monitored-battery = <&battery>;
+};
+
 &pmi8998_fg {
 	status = "okay";
 	monitored-battery = <&battery>;
@@ -415,11 +728,47 @@
 
 &pmi8998_rradc {
 	status = "okay";
+	monitored-battery = <&battery>;
 };
 
-&pmi8998_charger {
+
+&pmi8998_gpio {
+	pwm_fan_ctrl: pwm-fan-ctrl {
+		bias-disable;
+		drive-push-pull;
+		function = PMIC_GPIO_FUNC_FUNC1;
+		output-low;
+		pins = "gpio5";
+		power-source = <0>;
+		qcom,drive-strength = <PMIC_GPIO_STRENGTH_LOW>;
+	};
+};
+
+&pmi8998_lpg {
 	status = "okay";
-	monitored-battery = <&battery>;
+
+	multi-led {
+		color = <LED_COLOR_ID_RGB>;
+		function = LED_FUNCTION_STATUS;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		led@3 {
+			reg = <3>;
+			color = <LED_COLOR_ID_BLUE>;
+		};
+
+		led@4 {
+			reg = <4>;
+			color = <LED_COLOR_ID_GREEN>;
+		};
+
+		led@5 {
+			reg = <5>;
+			color = <LED_COLOR_ID_RED>;
+		};
+	};
 };
 
 &pmi8998_wled {
@@ -428,14 +777,6 @@
 	qcom,num-strings = <2>;
 };
 
-/* QUAT I2S Uses 1 I2S SD Line for audio on HDMI Bridge */
-&q6afedai {
-	qi2s@22 {
-		reg = <22>;
-		qcom,sd-lines = <0>;
-	};
-};
-
 &q6asmdai {
 	dai@0 {
 		reg = <0>;
@@ -508,22 +849,18 @@
 	cd-gpios = <&tlmm 126 GPIO_ACTIVE_LOW>;
 };
 
-&slpi_pas {
-	status = "okay";
-	firmware-name = "qcom/sdm845/odin/slpi.mbn";
-};
-
 &sound {
-	compatible = "qcom,db845c-sndcard";
-	pinctrl-0 = <&quat_mi2s_active
-			&quat_mi2s_sd0_active>;
-	pinctrl-names = "default";
+	compatible = "qcom,sdm845-sndcard";
 	model = "Odin";
+
 	audio-routing =
 		"RX_BIAS", "MCLK",
-		"AMIC1", "MIC BIAS1",
 		"AMIC2", "MIC BIAS2",
-		"AMIC3", "MIC BIAS3";
+		"SpkrLeft IN", "SPK1 OUT",
+		"SpkrRight IN", "SPK2 OUT",
+		"MM_DL1",  "MultiMedia1 Playback",
+		"MM_DL3",  "MultiMedia3 Playback",
+		"MultiMedia2 Capture", "MM_UL2";
 
 	mm1-dai-link {
 		link-name = "MultiMedia1";
@@ -575,11 +912,24 @@
 			sound-dai = <&wcd9340 1>;
 		};
 	};
+
+	slim-wcd-dai-link {
+		link-name = "SLIM WCD Playback";
+		cpu {
+			sound-dai = <&q6afedai SLIMBUS_1_RX>;
+		};
+
+		platform {
+			sound-dai = <&q6routing>;
+		};
+
+		codec {
+			sound-dai =  <&wcd9340 2>;
+		};
+	};
 };
 
 &tlmm {
-	gpio-reserved-ranges = <0 4>, <81 4>;
-
 	sdc2_default_state: sdc2-default {
 		clk {
 			pins = "sdc2_clk";
@@ -607,9 +957,8 @@
 	};
 
 	wcd_intr_default: wcd_intr_default {
-		pins = <54>;
+		pins = "gpio54";
 		function = "gpio";
-
 		input-enable;
 		bias-pull-down;
 		drive-strength = <2>;
@@ -629,6 +978,34 @@
 		drive-strength = <16>;
 		bias-pull-up;
 	};
+
+	gpio_leds_default: gpio-leds-default {
+		pins = "gpio7", "gpio25", "gpio87", "gpio129";
+		function = "gpio";
+		drive-strength = <16>;
+		bias-disable;
+	};
+
+	btn_en_active: btn-en-active {
+		pins = "gpio88";
+		function = "gpio";
+		drive-strength = <8>;
+		bias-disable;
+		output-high;
+	};
+
+	btn_pins: btn-pins {
+		mux {
+			pins = "gpio0", "gpio3", "gpio22", "gpio24",
+			       "gpio26", "gpio40", "gpio41", "gpio49",
+			       "gpio52", "gpio91", "gpio92", "gpio95",
+			       "gpio96", "gpio124", "gpio128", "gpio116",
+			       "gpio122";
+			function = "gpio";
+			drive-strength = <16>;
+			bias-pull-up;
+		};
+	};
 };
 
 &uart6 {
@@ -641,6 +1018,7 @@
 		vddxo-supply = <&vreg_l7a_1p8>;
 		vddrf-supply = <&vreg_l17a_1p3>;
 		vddch0-supply = <&vreg_l25a_3p3>;
+		vddch1-supply = <&vreg_l23a_3p3>;
 		max-speed = <3200000>;
 	};
 };
@@ -695,7 +1073,7 @@
 
 &venus {
 	status = "okay";
-	firmware-name = "qcom/sdm845/odin/venus.mbn";
+	firmware-name = "qcom/venus-5.2/venus.mbn";
 };
 
 &wcd9340{
@@ -709,6 +1087,9 @@
 	vdd-tx-supply = <&vreg_s4a_1p8>;
 	vdd-rx-supply = <&vreg_s4a_1p8>;
 	vdd-io-supply = <&vreg_s4a_1p8>;
+	qcom,mbhc-buttons-vthreshold-microvolt = <75000 150000 237000 500000 500000 500000 500000 500000>;
+	qcom,mbhc-headset-vthreshold-microvolt = <1700000>;
+	qcom,mbhc-headphone-vthreshold-microvolt = <50000>;
 
 	swm: swm@c85 {
 		left_spkr: wsa8810-left{
@@ -738,6 +1119,7 @@
 	vdd-1.8-xo-supply = <&vreg_l7a_1p8>;
 	vdd-1.3-rfa-supply = <&vreg_l17a_1p3>;
 	vdd-3.3-ch0-supply = <&vreg_l25a_3p3>;
+	vdd-3.3-ch1-supply = <&vreg_l23a_3p3>;
 
 	qcom,snoc-host-cap-8bit-quirk;
 };
