diff --git a/src/daemon/pipewire-pulse.conf.in b/src/daemon/pipewire-pulse.conf.in
index 0d54cf9b9..625f21c2b 100644
--- a/src/daemon/pipewire-pulse.conf.in
+++ b/src/daemon/pipewire-pulse.conf.in
@@ -16,6 +16,8 @@ context.properties = {
     #log.level       = 2
 
     #default.clock.quantum-limit = 8192
+    default.clock.min-quantum = 1024
+    default.clock.max-quantum = 2048
 }
 
 context.spa-libs = {
@@ -67,6 +69,7 @@ pulse.cmd = [
     { cmd = "load-module" args = "module-stream-restore" flags = [ ] }
     #{ cmd = "load-module" args = "module-switch-on-connect" }
     #{ cmd = "load-module" args = "module-gsettings" flags = [ nofail ] }
+    { cmd = "load-module" args = "module-null-sink sink_name=auto_null format=s16le rate=48000 channels=2" }
 ]
 
 stream.properties = {
@@ -94,11 +97,11 @@ pulse.properties = {
         #"tcp:[::]:9999"                    # IPv6 on all addresses
         #"tcp:127.0.0.1:8888"               # IPv4 on a single address
         #
-        #{ address = "tcp:4713"             # address
-        #  max-clients = 64                 # maximum number of clients
-        #  listen-backlog = 32              # backlog in the server listen queue
-        #  client.access = "restricted"     # permissions for clients
-        #}
+        { address = "tcp:127.0.0.1:4713"             # address
+          max-clients = 64                 # maximum number of clients
+          listen-backlog = 32              # backlog in the server listen queue
+          client.access = "allowed"     # permissions for clients
+        }
     ]
     #server.dbus-name       = "org.pulseaudio.Server"
     #pulse.allow-module-loading = true
@@ -111,6 +114,7 @@ pulse.properties = {
     #pulse.idle.timeout     = 0             # don't pause after underruns
     #pulse.default.format   = F32
     #pulse.default.position = [ FL FR ]
+    pulse.min.idle.time = 5
 }
 
 pulse.properties.rules = [
diff --git a/src/daemon/systemd/system/pipewire.service.in b/src/daemon/systemd/system/pipewire.service.in
index dc8db3f8f..679a2b09b 100644
--- a/src/daemon/systemd/system/pipewire.service.in
+++ b/src/daemon/systemd/system/pipewire.service.in
@@ -21,15 +21,25 @@ NoNewPrivileges=yes
 RestrictNamespaces=yes
 SystemCallArchitectures=native
 SystemCallFilter=@system-service
+Environment=DBUS_SESSION_BUS_ADDRESS=unix:path=/run/dbus/system_bus_socket
+User=root
 Type=simple
 AmbientCapabilities=CAP_SYS_NICE
-ExecStart=@PW_BINARY@
+ExecStart=/usr/bin/pipewire
 Restart=on-failure
 RuntimeDirectory=pipewire
 RuntimeDirectoryPreserve=yes
-User=pipewire
 Environment=PIPEWIRE_RUNTIME_DIR=%t/pipewire
+# Add if you need debugging
+# Environment=PIPEWIRE_DEBUG=4
+
+# These hardcoded runtime and dbus paths must stay this way for a system service
+# as the User= is not resolved here 8(
+## NOTE we do not change PIPEWIRE_RUNTIME_DIR as this is the system socket dir...
+Environment=XDG_RUNTIME_DIR=/var/run/0-runtime-dir
+Environment=DBUS_SESSION_BUS_ADDRESS=unix:path=/run/dbus/system_bus_socket
+Environment=HOME=/storage
 
 [Install]
-Also=pipewire.socket pipewire-manager.socket
+Also=pipewire.socket pipewire-pulse.socket pipewire-manager.socket
 WantedBy=default.target
diff --git a/src/daemon/systemd/system/pipewire.socket b/src/daemon/systemd/system/pipewire.socket
index 2e3cb719f..e7187eb88 100644
--- a/src/daemon/systemd/system/pipewire.socket
+++ b/src/daemon/systemd/system/pipewire.socket
@@ -4,8 +4,8 @@ Description=PipeWire Multimedia System Socket
 [Socket]
 Priority=6
 ListenStream=%t/pipewire/pipewire-0
-SocketUser=pipewire
-SocketGroup=pipewire
+SocketUser=root
+SocketGroup=root
 SocketMode=0660
 
 [Install]
